File: shuffle-service/src/test/scala/org/apache/spark/shuffle/ytsaurus/YTsaurusShuffleManagerTest.scala
package org.apache.spark.shuffle.ytsaurus

import org.apache.spark.executor.TaskMetrics
import org.apache.spark.memory.TaskMemoryManager
import org.apache.spark.shuffle.{ShuffleHandle, ShuffleManager}
import org.apache.spark.{Partitioner, ShuffleDependency, SparkConf, SparkContext, TaskContext}
import org.mockito.Mockito
import org.mockito.scalatest.MockitoSugar
import org.scalatest.BeforeAndAfterEach
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers
import tech.ytsaurus.spyt.test.LocalYt

class YTsaurusShuffleManagerTest extends AnyFlatSpec with Matchers with BeforeAndAfterEach with MockitoSugar {
  behavior of "YTsaurusShuffleManager"

  private var sc: SparkContext = _
  private var shuffleManager: ShuffleManager = _
  private var handle: ShuffleHandle = _

  override protected def beforeEach(): Unit = {
    super.beforeEach()
    shuffleManager = null
    handle = null
  }

  override protected def afterEach(): Unit = {
    super.afterEach()
    if (sc != null) {
      sc.stop()
    }
    if (shuffleManager != null && handle != null) {
      shuffleManager.unregisterShuffle(handle.shuffleId)
    }
  }

  private def createTaskContextStub(partitionId: Int): TaskContext = {
    val taskContext = mock[TaskContext](Mockito.withSettings().lenient())
    val metrics = TaskMetrics.empty
    when(taskContext.taskMetrics()).thenReturn(metrics)
    when(taskContext.partitionId()).thenReturn(partitionId)
    Mockito.doNothing().when(taskContext).killTaskIfInterrupted()
    taskContext
  }

  private def testTemplate(
    shuffleId: Int,
    nMapTasks: Int,
    nOutputPartitions: Int,
    extraConf: Map[String, String] = Map.empty,
    skipOutputPartitions: Set[Int] = Set.empty
  ): Unit = {
    val conf = new SparkConf()
    conf
      .set("spark.shuffle.manager", "org.apache.spark.shuffle.ytsaurus.YTsaurusShuffleManager")
      .set("spark.shuffle.sort.io.plugin.class", "tech.ytsaurus.spyt.shuffle.YTsaurusShuffleDataIO")
      .set("spark.hadoop.yt.proxy", LocalYt.proxy)
      .set("spark.hadoop.yt.user", "root")
      .set("spark.hadoop.yt.token", "")
      .set("spark.ytsaurus.shuffle.replication.factor", "1")

    extraConf.foreach { case (key, value) => conf.set(key, value) }

    sc = new SparkContext("local", "test", conf)
    val shuffleManager = sc.env.shuffleManager

    shuffleManager shouldBe a[YTsaurusShuffleManager]

    val shuffleMapRdd = sc.parallelize(1 to 100).map(x => x -> s"Value $x")
    val shuffleDependency = new ShuffleDependency(shuffleMapRdd, new IntPartitioner(nOutputPartitions))
    handle = shuffleManager.registerShuffle(shuffleId, shuffleDependency)

    (0 until nMapTasks).foreach { mapId =>
      val taskContext = createTaskContextStub(mapId)
      val metrics = taskContext.taskMetrics.shuffleWriteMetrics
      val dataToWrite = (0 until (nOutputPartitions*3))
        .map(pId => (pId / 3) -> s"M$mapId P$pId")
        .filter(row => !skipOutputPartitions.contains(row._1))

      val writer = shuffleManager.getWriter[Int, String](handle, mapId, taskContext, metrics)
      writer.write(dataToWrite.iterator)
      writer.stop(success = true)
    }

    (0 until nOutputPartitions).foreach { reduceId =>
      val taskContext = createTaskContextStub(reduceId)
      val metrics = taskContext.taskMetrics.createTempShuffleReadMetrics()
      val reader = shuffleManager.getReader[Int, String](handle, reduceId, reduceId + 1, taskContext, metrics)
      val result = reader.read().toList
      val expected = if (skipOutputPartitions.contains(reduceId)) Nil else (0 until nMapTasks).flatMap { mapId =>
        (0 until 3).map(itemId => reduceId -> s"M$mapId P${reduceId*3 + itemId}")
      }

      result should contain theSameElementsAs expected
    }

  }

  it should "write and read shuffle data" in {
    testTemplate(1, 4, 10)
  }

  it should "read empty shuffle partitions without compression" in {
    testTemplate(1, 4, 10, Map("spark.shuffle.compress"-> "false"), skipOutputPartitions = Set(3,5))
  }

  it should "read empty shuffle partitions with compression" in {
    testTemplate(1, 4, 10, Map("spark.shuffle.compress"-> "true"), skipOutputPartitions = Set(3,5))
  }
}

class IntPartitioner(override val numPartitions: Int) extends Partitioner {
  override def getPartition(key: Any): Int = key.asInstanceOf[Int]
}

File: shuffle-service/src/test/java/org/apache/spark/shuffle/ytsaurus/YTsaurusShuffleManagerTest.java
package org.apache.spark.shuffle.ytsaurus;

import org.apache.spark.Partitioner;
import org.apache.spark.ShuffleDependency;
import org.apache.spark.SparkConf;
import org.apache.spark.SparkEnv;
import org.apache.spark.TaskContext;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.executor.TaskMetrics;
import org.apache.spark.shuffle.ShuffleHandle;
import org.apache.spark.shuffle.ShuffleManager;
import org.apache.spark.shuffle.ShuffleWriter;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import tech.ytsaurus.spyt.test.LocalYt;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;

import scala.Product2;
import scala.Tuple2;
import scala.collection.JavaConverters;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

class YTsaurusShuffleManagerTest {
    private JavaSparkContext sc;
    private ShuffleManager shuffleManager;
    private ShuffleHandle handle;

    @AfterEach
    void tearDown() {
        try {
            if (shuffleManager != null && handle != null) {
                shuffleManager.unregisterShuffle(handle.shuffleId());
            }
        } finally {
            if (sc != null) {
                sc.close();
            }
        }
    }

    private TaskContext taskContext(int partitionId) {
        TaskContext context = mock(TaskContext.class);
        when(context.taskMetrics()).thenReturn(TaskMetrics.empty());
        when(context.partitionId()).thenReturn(partitionId);
        return context;
    }

    @SuppressWarnings("unchecked")
    private void testShuffle(Map<String, String> extraConf, Set<Integer> skippedPartitions) throws Exception {
        int shuffleId = 1;
        int mapTasks = 4;
        int outputPartitions = 10;
        SparkConf conf = new SparkConf(false)
                .setMaster("local")
                .setAppName("test")
                .set("spark.shuffle.manager", "org.apache.spark.shuffle.ytsaurus.YTsaurusShuffleManager")
                .set("spark.shuffle.sort.io.plugin.class", "tech.ytsaurus.spyt.shuffle.YTsaurusShuffleDataIO")
                .set("spark.hadoop.yt.proxy", LocalYt.proxy())
                .set("spark.hadoop.yt.user", "root")
                .set("spark.hadoop.yt.token", "")
                .set("spark.ytsaurus.shuffle.replication.factor", "1");
        extraConf.forEach(conf::set);
        sc = new JavaSparkContext(conf);
        shuffleManager = SparkEnv.get().shuffleManager();
        assertInstanceOf(YTsaurusShuffleManager.class, shuffleManager);

        var rdd = sc.parallelizePairs(Collections.singletonList(new Tuple2<>(1, "Value 1")))
                .partitionBy(new IntPartitioner(outputPartitions));
        ShuffleDependency<Integer, String, String> dependency =
                (ShuffleDependency<Integer, String, String>) rdd.rdd().dependencies().head();
        handle = shuffleManager.registerShuffle(shuffleId, dependency);

        for (int mapId = 0; mapId < mapTasks; mapId++) {
            TaskContext context = taskContext(mapId);
            List<Product2<Integer, String>> data = new ArrayList<>();
            for (int pId = 0; pId < outputPartitions * 3; pId++) {
                if (!skippedPartitions.contains(pId / 3)) {
                    data.add(new Tuple2<>(pId / 3, "M" + mapId + " P" + pId));
                }
            }
            ShuffleWriter<Integer, String> writer = shuffleManager.getWriter(
                    handle, mapId, context, context.taskMetrics().shuffleWriteMetrics());
            writer.write(JavaConverters.asScalaIteratorConverter(data.iterator()).asScala());
            assertNull(YTsaurusShuffleExecutorComponents.currentHandle().get());
            assertNull(YTsaurusShuffleExecutorComponents.currentMapIndex().get());
            writer.stop(true);
        }

        for (int reduceId = 0; reduceId < outputPartitions; reduceId++) {
            TaskContext context = taskContext(reduceId);
            var metrics = context.taskMetrics().createTempShuffleReadMetrics();
            var reader = shuffleManager.<Integer, String>getReader(handle, reduceId, reduceId + 1, context, metrics);
            List<Tuple2<Integer, String>> result = new ArrayList<>();
            var records = reader.read();
            while (records.hasNext()) {
                Product2<Integer, String> record = records.next();
                result.add(new Tuple2<>(record._1(), record._2()));
            }
            List<Tuple2<Integer, String>> expected = new ArrayList<>();
            if (!skippedPartitions.contains(reduceId)) {
                for (int mapId = 0; mapId < mapTasks; mapId++) {
                    for (int itemId = 0; itemId < 3; itemId++) {
                        expected.add(new Tuple2<>(reduceId, "M" + mapId + " P" + (reduceId * 3 + itemId)));
                    }
                }
            }
            Comparator<Tuple2<Integer, String>> order = Comparator.comparing(Tuple2::_2);
            result.sort(order);
            expected.sort(order);
            assertEquals(expected, result);
            assertEquals(expected.size(), metrics.recordsRead());
        }
    }

    @Test
    void readsEmptyPartitionsWithoutCompression() throws Exception {
        testShuffle(Map.of("spark.shuffle.compress", "false"), Set.of(3, 5));
    }

    @Test
    void readsEmptyPartitionsWithCompression() throws Exception {
        testShuffle(Map.of("spark.shuffle.compress", "true"), Set.of(3, 5));
    }

    @Test
    void writesAndReadsPushBasedShuffle() throws Exception {
        testShuffle(Map.of("spark.ytsaurus.shuffle.push.enabled", "true"), Set.of(3, 5));
    }

    private static class IntPartitioner extends Partitioner {
        private final int numPartitions;

        IntPartitioner(int numPartitions) {
            this.numPartitions = numPartitions;
        }

        @Override
        public int numPartitions() {
            return numPartitions;
        }

        @Override
        public int getPartition(Object key) {
            return (Integer) key;
        }
    }
}
